### PR TITLE
Backport 9.4 create/drop extension feature

### PR DESCRIPTION
**Pull request by funnyxj**
I backport pg9.4 extension feature, Core support for "extensions", which are packages of SQL objects.Howere alter extension is not supported, but I can do it later.

CREATE EXTENSION
Name

CREATE EXTENSION -- install an extension
Synopsis

CREATE EXTENSION [ IF NOT EXISTS ] extension_name
    [ WITH ] [ SCHEMA schema_name ]
             [ VERSION version ]
             [ FROM old_version ]
Description

CREATE EXTENSION loads a new extension into the current database. There must not be an extension of the same name already loaded.

Loading an extension essentially amounts to running the extension's script file. The script will typically create new SQL objects such as functions, data types, operators and index support methods. CREATE EXTENSION additionally records the identities of all the created objects, so that they can be dropped again if DROP EXTENSION is issued.

Loading an extension requires the same privileges that would be required to create its component objects. For most extensions this means superuser or database owner privileges are needed. The user who runs CREATE EXTENSION becomes the owner of the extension for purposes of later privilege checks, as well as the owner of any objects created by the extension's script.

Parameters

IF NOT EXISTS
Do not throw an error if an extension with the same name already exists. A notice is issued in this case. Note that there is no guarantee that the existing extension is anything like the one that would have been created from the currently-available script file.

extension_name
The name of the extension to be installed. PostgreSQL will create the extension using details from the file SHAREDIR/extension/extension_name.control.

schema_name
The name of the schema in which to install the extension's objects, given that the extension allows its contents to be relocated. The named schema must already exist. If not specified, and the extension's control file does not specify a schema either, the current default object creation schema is used.

Remember that the extension itself is not considered to be within any schema: extensions have unqualified names that must be unique database-wide. But objects belonging to the extension can be within schemas.

version
The version of the extension to install. This can be written as either an identifier or a string literal. The default version is whatever is specified in the extension's control file.

old_version
FROM old_version must be specified when, and only when, you are attempting to install an extension that replaces an "old style" module that is just a collection of objects not packaged into an extension. This option causes CREATE EXTENSION to run an alternative installation script that absorbs the existing objects into the extension, instead of creating new objects. Be careful that SCHEMA specifies the schema containing these pre-existing objects.

The value to use for old_version is determined by the extension's author, and might vary if there is more than one version of the old-style module that can be upgraded into an extension. For the standard additional modules supplied with pre-9.1 PostgreSQL, use unpackaged for old_version when updating a module to extension style.

### CODE REVIEW

**Comment by funnyxj** [comment]
@danielgustafsson  Can you review this patch?

**Comment by danielgustafsson** [comment]
Hi @funnyxj,

First of all, thanks for your patch! Of course I will review it. As the bot commented, we have a [CLA](http://greenplum.org/contributor_agreements/GPDB_Pivotal_ICLA.pdf) for code submissions, please have a look and sign it so that code can be pushed to the repo.

This is a large body of work and it will take some time to review so bear with me. Are you subscribed to the developer mailinglist in case any part of this gets discussed there?

**Comment by ivannovick** [comment]
This is great work!!!!  Let me know if you need help getting more reviewers i am in favor of getting this merged after it passes code review

**Comment by ascherbaum-pivotal** [comment]
That is indeed a very valuable contribution, thank you!

How does this play together with gppkg, or how are cluster expansions handled?

**Comment by yydzero** [comment]
gppkg is a collection of rpms like postgis, plpython etc, so should not
block this PR. we could tackle it in later commits.

On Wed, Jun 15, 2016 at 12:06 AM, Andreas Scherbaum <
notifications@github.com> wrote:

> That is indeed a very valuable contribution, thank you!
>
> How does this play together with gppkg, or how are cluster expansions
> handled?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/835#issuecomment-225931264>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAfRk1vjf5BY-B9oyWmYrHcW1hihydNOks5qLtGVgaJpZM4I0Hkk>
> .
>



-- 
Best Regards,
Yandong

File: src/backend/commands/extension.c [hunk 1]
```diff
@@ -0,0 +1,1748 @@
...
+#include "utils/tqual.h"
+
+#include "cdb/cdbvars.h"
+#include "cdb/cdbdisp.h"
+
+/* Globally visible state variables */
+bool		creating_extension = false;
+
+/* Knob to control if we treat the invoking user of CREATE/ALTER EXTENSION as super user */
+bool		rds_enable_promoting_privilege = true;
```
Comment (yydzero) [inline src/backend/commands/extension.c]: Is rds_enable_promoting_privilege used anywhere? could not find any reference

**Comment by lij55** [comment]
This is really great feature! Thanks, funnyxy!

One comment, is it possible to provide some tests that cover the features simultaneously?

File: src/backend/commands/extension.c (hunk 1 above)
Comment (funnyxj) [inline src/backend/commands/extension.c]: Thank you，I will delete it

**Comment by funnyxj** [comment]
I prepare to submit  CLA @danielgustafsson .

I will  provide some tests that cover the features @lij55

**Comment by funnyxj** [comment]
gppkg is not conflict with extension, @ascherbaum-pivotal

**Comment by ascherbaum-pivotal** [comment]
Looking through the changes, I miss support for dump & restore.

**Comment by ascherbaum-pivotal** [comment]
Do you plan to add the system views "pg_available_extensions" and "pg_available_extension_versions"?

**Comment by hlinnaka** [comment]
Thanks @funnyxj!

One thing caught my eye: you replaced the upstream SysCache lookup calls with caql calls. That's understandable, since we had done that all over the greenplum codebase earlier, but these days the whole caql thing is deprecated, and there's no need to do that in new code. So, it'd be better to just keep those the way they are in the upstream.

(Mea culpa, that's not mentioned in any docs. And we really should bite the bullet and search&replace all those caql calls back to SysCache calls, and remove the caql code completely, to avoid the confusion..)

**Comment by danielgustafsson** [comment]
Just to give a status message, I am reviewing this patch rebased on top of the 8.3 merged tree and will get back to you shortly with some feedback.

**Comment by hlinnaka** [comment]
I rebased this over current master, cleaned up a bunch of issues:

* changed some whitespace, variable names, and moved some functions to where they are in the upstream source files, etc. to match the upstream version more closely. I used PostgreSQL 9.1 stable as the reference point, as that's the first version that had extensions, and that's what we're going to encounter first as we merge the whole codebase with upstream.

* Removed a few functions that were unused

* Replaced the caql calls with upstream syscache lookups

See https://github.com/hlinnaka/gpdb/tree/feature_extension_patch for those changes.

There are some remaining issues that will need to be resolved:

* There are no regression tests. The upstream didn't have any explicit regression tests for the new commands either, but all the contrib modules were turned into extensions, so we got testing that way. I think we need to do the same in GPDB. And perhaps add some explicit regression tests too, just to be sure.

* I think we'll need the ALTER commands too. And as @ascherbaum-pivotal noted, the system views are missing.

* Does this make sure the OID of the extension is the same in the master and the QE nodes? I think CreateExtension struct needs to be modified to include the OID, and the master needs to fill it in when it creates the pg_extension row. Most DDL commands in GPDB have been modified that way. For example, take a look at commit 604bb131ffbc94f790f6e8afafa8e30294e4a4b5, which recently did that for the text search objects, which we forgot about in the 8.3 merge.

**Comment by hlinnaka** [comment]
Oh, also, pg_dump/restore support is missing.

**Comment by yydzero** [comment]
@hlinnaka how about we merge this commit after make build pass and then handle remaining issues in coming commits?

**Comment by hlinnaka** [comment]
On 13 October 2016 17:54:48 EEST, Yandong Yao <notifications@github.com> wrote:
>@hlinnaka how about we merge this commit after make build pass and then
>handle remaining issues in coming commits?

I would recommend against that.

- Heikki

**Comment by yydzero** [comment]
ok, I am working on a new PR based on above efforts, will resolve all above comments.

**Comment by ivannovick** [comment]
thx @yydzero !!!

**Comment by yydzero** [comment]
replace this with #1266

**Pull request closed without merging**